app/api/scorecard.py

```diff
@@ -0,0 +1,140 @@
+from app import dependencies
+from fastapi import APIRouter, Depends
+from fastapi.encoders import jsonable_encoder
+from app import schemas, crud
+from typing import Any, List
+from app.utils.base import format_code, check_metric, stringify_value
+from .exceptions import HTTPResponseCustomized
+from .responses import ResponseCustomized
+from fastapi.responses import ORJSONResponse
+from app.schemas.apiResponse import CustomResponse
+from app.schemas.scorecardServiceMetric import ScorecardServiceMetricCreate
+from app.crud.scorecardServiceMetric import CRUDScoreCardServiceMetric
+
+
+router = APIRouter()
+@router.post("/", response_model=CustomResponse, response_class=ResponseCustomized)
+def createScoreCard(scoreCardInput: ScorecardServiceMetricCreate,
+                    scoreCardCrud: crud.CRUDScoreCard = Depends(
+                        dependencies.getScoreCardsCrud),
+                    serviceScorecardCrud: crud.CRUDMicroserviceScoreCard = Depends(
+                        dependencies.getMicroserviceScoreCardsCrud),
+                    scorecardMetricCrud: crud.CRUDScoreCardMetric = Depends(
+                        dependencies.getScoreCardMetricsCrud),
+                    serviceCrud: crud.CRUDMicroservice = Depends(
+                        dependencies.getMicroservicesCrud),
+                    metricCrud: crud.CRUDMetric = Depends(
+                        dependencies.getMetricsCrud),
+                    scorecard: crud.CRUDScoreCard = Depends(
+                        dependencies.getScoreCardsCrud)
+                    ):
+    # Create the scorecard.
+    scoreCardInput.code = format_code(scoreCardInput.name)
+    scorecard = schemas.ScoreCardCreate(
+        name=scoreCardInput.name,
+        code=scoreCardInput.code,
+        description=scoreCardInput.description
+    )
+    if (scoreCardCrud.getByScoreCardCode(scoreCardInput.code)):
+        raise HTTPResponseCustomized(
+            status_code=400, detail="Scorecard already found")
+    try:
+        obj = scoreCardCrud.create(scorecard)
+        scorecardID = obj.id
+    except Exception as e:
+        raise HTTPResponseCustomized(status_code=422, detail=f"Error: {e}")
+
+    # if the scorecardids are passed duplicated
+    servicesIDsSet = list(set(scoreCardInput.services))
+    # i need to check if the serviceid is found or not
+    service = serviceCrud.getByServiceIds(servicesIDsSet)
+    serviceListToBeCreated = []
+    for serviceid in servicesIDsSet:
+        if (len(service) == len(servicesIDsSet)):
```

Comment:
the condition shouldn't be inside the for loop.
the condition check whether the length of the list of retrieved services from the DB is equal the length of the passed service_ids, if they are not equal this means that the passed service_ids list contains invalid data so you can abort the request otherwise continue without running the check again.

Reply:
the check should be before creating the scorecard object, such that if the check failed you can raise HTTPException before mutating the DB

Reply:
you can also move the check on metric_ids right after the check on service_ids, and in case any of the two checks fails raise the exception before mutating the DB.